### PR TITLE
Fixes #16167 - added list of safemode methods

### PR DESCRIPTION
**Pull request by lzap**
Visit About page to see the list. Could not figure out better place for this.
Maybe template editor (as hidden box). It is worth to cache this table perhaps
too.

### CODE REVIEW

File: app/controllers/about_controller.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ def index
     @smart_proxies = SmartProxy.authorized(:view_smart_proxies).includes(:features)
     @compute_resources = ComputeResource.authorized(:view_compute_resources)
     @plugins = Foreman::Plugin.all
+    @safemode = ObjectSpace.each_object(Class).select{|x| x < Safemode::Jail }.collect{|x| [x.name.gsub(/::Jail$/, ''), x.allowed_methods.sort.join(' ')]}.sort{|a,b| a.first <=> b.first}
```
Comment (domcleal) [inline app/controllers/about_controller.rb]: Please address the Code Climate result, this is very complex. It should be moved into a helper at the very least, and split into multiple lines with meaningful variable names.

**Comment by domcleal** [comment]
> Maybe template editor (as hidden box).

This seems eminently more sensible than the About page, can you do that?

It's also missing tests.

**Comment by lzap** [comment]
Done moving, I also moved the text field (the editor) under the name, it required scrolling on smaller screens:

image

If you click on the info box link, it will switch tabs to:

image

Note this is only available when safemode is turned on.

File: app/helpers/templates_helper.rb [hunk 2]
```diff
@@ -28,6 +28,20 @@ def infobox_functions_and_macros(docs_section)
           :rel => 'external')
   end
 
+  def safemode_methods
+    Rails.cache.fetch("safemode_methods", :expires_in => 2.hours) do
+      objects = ObjectSpace.each_object(Class).select{|x| x < Safemode::Jail }
+      objects_with_methods = objects.collect do |obj|
```
[the remark this answers is not in the document]
Comment (lzap) [reply app/helpers/templates_helper.rb]: Changed.

File: app/helpers/templates_helper.rb [hunk 3]
```diff
@@ -28,6 +28,20 @@ def infobox_functions_and_macros(docs_section)
           :rel => 'external')
   end
 
+  def safemode_methods
+    Rails.cache.fetch("safemode_methods", :expires_in => 2.hours) do
```
Comment (domcleal) [inline app/helpers/templates_helper.rb]: The version number (`SETTING[:version]`) should probably be added to the cache key, then it'll automatically invalidate on update.

File: app/helpers/templates_helper.rb [hunk 4]
```diff
@@ -28,6 +28,20 @@ def infobox_functions_and_macros(docs_section)
...
+      objects = ObjectSpace.each_object(Class).select{|x| x < Safemode::Jail }
+      objects_with_methods = objects.map do |obj|
+        [obj.name.gsub(/::Jail$/, ''), obj.allowed_methods.sort.join(' ')]
+      end
+      objects_with_methods.uniq.sort{|a,b| a.first <=> b.first}
```
Comment (domcleal) [inline app/helpers/templates_helper.rb]: `sort_by(&:first)`

File: app/helpers/templates_helper.rb [hunk 5]
```diff
@@ -28,6 +28,20 @@ def infobox_functions_and_macros(docs_section)
...
+    end
+  end
+
+  def infobox_safemode
+    alert(:class => 'alert-info', :header => '', :text => (_('In safemode only %s are allowed') % link_to_function(_('selected methods or variables'), '$("#safemode_tab").click()')))
```
Comment (domcleal) [inline app/helpers/templates_helper.rb]: a) I'm unsure if "safemode" is the best phrase to use.  I'd suggest keeping it really simple and use "In templates", because you already know safemode is enabled.  Else I'd consider calling it "safe mode", as this is the description used in settings - the name of the gem isn't the user's business.
b) Avoid splitting the sentence into three parts over two strings, they won't be translated together and translators won't know they're from the same text. I'd suggest having one sentence, a colon, then a link that can be translated more independently (e.g. "list of available template helpers").

File: app/views/about/index.html.erb [hunk 6]
```diff
@@ -13,6 +13,7 @@
           <li><a href="#compute_resources" data-toggle="tab"><%= _('Compute Resources') %></a></li>
         <% end %>
         <li><a href="#plugins" data-toggle="tab"><%= _('Plugins') %></a></li>
+        <li><a href="#safemode" data-toggle="tab"><%= _('Safemode') %></a></li>
```
Comment (domcleal) [inline app/views/about/index.html.erb]: ditto about "Safemode", consider making this more generic - "Template help", "Help" or similar?

File: app/views/provisioning_templates/_alerts.html.erb [hunk 7]
```diff
@@ -1,3 +1,5 @@
+<%= infobox_safemode if Setting[:safemode_render] %>
```
Comment (domcleal) [inline app/views/provisioning_templates/_alerts.html.erb]: Should this be in the `else` branch with the other help box?

File: app/views/templates/_form.html.erb [hunk 8]
```diff
@@ -9,35 +9,22 @@
...
-
     <li><a id='history_tab' href="#history" data-toggle="tab"><%= _("History") %></a></li>
     <% if show_location_tab? %>
       <li><a href="#locations" data-toggle="tab"><%= _("Locations") %></a></li>
     <% end %>
     <% if show_organization_tab? %>
       <li><a href="#organizations" data-toggle="tab"><%= _("Organizations") %></a></li>
     <% end %>
+    <% if Setting[:safemode_render] %>
+      <li><a id='safemode_tab' href="#safemode" data-toggle="tab"><%= _("Safemode") %></a></li>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: ditto re naming

File: app/views/templates/_form.html.erb [hunk 9]
```diff
@@ -9,35 +9,22 @@
...
 
   <div class="tab-content">
     <div class="tab-pane active" id="primary">
       <%= text_f f, :name, :disabled => @template.locked? %>
 
-      <% if show_default? %>
-        <%= checkbox_f f, :default, :label=>_('Default'), :help_block => default_template_description %>
-      <% end -%>
-
-      <%= render "custom_tabs", :f => f if type == 'ptable' %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb:30]: Why move the custom tabs and default checkbox? This is unrelated to the alerts, and spreads the ptable UI about somewhat.

File: app/views/templates/_form.html.erb [hunk 10]
```diff
@@ -84,6 +85,32 @@
...
+        <thead>
+          <tr>
+            <th><%= _("Class") %></th>
+            <th><%= _("Allowed methods or members") %></th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr>
+            <td class="ellipsis"><em><%= _("Global methods (functions)") %></em></td>
+            <td><%= Foreman::Renderer::ALLOWED_HELPERS.join(' ') %></td>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: This column should all have code style formatting, not plain text.

File: app/views/templates/_form.html.erb [hunk 11]
```diff
@@ -84,6 +85,32 @@
...
+          </tr>
+          <tr>
+            <td class="ellipsis"><em><%= _("Global variables") %></em></td>
+            <td><%= Foreman::Renderer::ALLOWED_VARIABLES.map{|x| "@#{x}"}.join(' ') %></td>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: The list of variables (and helpers, above) should be sorted to match the jail methods.

File: app/helpers/templates_helper.rb [hunk 12]
```diff
@@ -28,6 +28,20 @@ def infobox_functions_and_macros(docs_section)
           :rel => 'external')
   end
 
+  def safemode_methods
```
Comment (domcleal) [inline app/helpers/templates_helper.rb:23]: Please add a unit test for this method under test/unit/helpers/ to check it returns some known jail members.

File: app/helpers/templates_helper.rb (hunk 3 above)
Comment (lzap) [inline app/helpers/templates_helper.rb]: > The version number (`SETTING[:version]`) should probably be added to the cache key, then it'll automatically invalidate on update.

Good idea, do we also have unique UUID of a process? I could simply use
PID for that (so it invalidates after restart). Cos only restart can
add/remove from the list. Then I could set expire to never.

## 

Later,
 Lukas #lzap Zapletal

Comment (domcleal) [reply app/helpers/templates_helper.rb]: No point in using the Rails cache if you intend to create a per-process cache, use a class variable instead.

Comment (lzap) [reply app/helpers/templates_helper.rb]: Indeed, missed that option completely. Will use that instead.

File: app/helpers/templates_helper.rb (hunk 5 above)
Comment (lzap) [reply app/helpers/templates_helper.rb]: Sure, going for: _In templates only selected methods and variables are allowed, list of available template helpers_

File: app/views/about/index.html.erb (hunk 6 above)
Comment (lzap) [reply app/views/about/index.html.erb]: Good idea, I like "Help"

File: app/views/provisioning_templates/_alerts.html.erb (hunk 7 above)
Comment (lzap) [reply app/views/provisioning_templates/_alerts.html.erb]: Absolutely.

File: app/views/templates/_form.html.erb (hunk 9 above)
Comment (lzap) [reply app/views/templates/_form.html.erb:30]: Because I think it's better UX, but I admit it does not belong in here.

File: app/views/templates/_form.html.erb (hunk 10 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Tried `<code>` but it looked ugly, created a CSS class for that.

File: app/views/templates/_form.html.erb (hunk 11 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Yep, extracted it into helpers as well.

File: app/helpers/templates_helper.rb (hunk 12 above)
Comment (lzap) [reply app/helpers/templates_helper.rb:23]: I think that's overtesting, covered with functional test. But I will add one.

**Comment by lzap** [comment]
https://gist.github.com/e843217de35a543fb417b48c66451f5c

File: app/views/about/index.html.erb (hunk 6 above)
Comment (domcleal) [reply app/views/about/index.html.erb]: This tab is unused?

File: app/helpers/templates_helper.rb [hunk 13]
```diff
@@ -28,6 +28,28 @@ def infobox_functions_and_macros(docs_section)
...
+  def safemode_helpers
+    @@safemode_helpers ||= Foreman::Renderer::ALLOWED_HELPERS.sort.join(' ')
+  end
+
+  def safemode_variables
+    @@safemode_variables ||= Foreman::Renderer::ALLOWED_VARIABLES.sort.map{|x| "@#{x}"}.join(' ')
+  end
+
+  def infobox_safemode
+    alert(:class => 'alert-info', :header => '', :text => (_('In templates only selected methods and variables are allowed') + ', ' + link_to_function(_('list of available template helpers'), '$("#help_tab").click()')))
```
Comment (domcleal) [inline app/helpers/templates_helper.rb]: I don't think this alert is really necessary. The "Help" tab is quite clear and visible, so this only uses up screen estate, I'd just remove it.

File: app/views/provisioning_templates/_alerts.html.erb [hunk 14]
```diff
@@ -1,5 +1,6 @@
 <% if @template.locked? -%>
   <%= locked_warning(::Template.find(params[:id])) %>
 <% else -%>
+  <%= infobox_safemode if Setting[:safemode_render] %>
   <%= infobox_functions_and_macros("4.4.3ProvisioningTemplates") %>
```
Comment (domcleal) [inline app/views/provisioning_templates/_alerts.html.erb]: This can move onto the Help tab, either as a block of text or alert at the top of the tab, it'd be a much better place for it.

File: app/views/templates/_form.html.erb [hunk 15]
```diff
@@ -9,33 +9,26 @@
...
     <%= render "#{@type_name_plural}/custom_tab_headers" unless type == 'ptable'%>
-
     <li><a id='history_tab' href="#history" data-toggle="tab"><%= _("History") %></a></li>
     <% if show_location_tab? %>
       <li><a href="#locations" data-toggle="tab"><%= _("Locations") %></a></li>
     <% end %>
     <% if show_organization_tab? %>
       <li><a href="#organizations" data-toggle="tab"><%= _("Organizations") %></a></li>
     <% end %>
+    <% if Setting[:safemode_render] %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: This will need to be always displayed when moving the link to docs.

File: app/assets/stylesheets/application.scss [hunk 16]
```diff
@@ -573,3 +573,7 @@ table {
 .masked-input {
   font-family: 'AllBullets';
 }
+
+.code-in-table {
```
Comment (domcleal) [inline app/assets/stylesheets/application.scss]: There should already be styling of `<code>` tags from Bootstrap, is this needed? (and should it be `table code`?)

File: app/assets/stylesheets/application.scss [hunk 17]
```diff
@@ -573,3 +573,7 @@ table {
...
+  font-family: DejaVu Sans Mono, Courier New, monospace;
```
Comment (domcleal) [inline app/assets/stylesheets/application.scss]: Not the standard Bootstrap font family for code blocks, why change it?

File: app/assets/stylesheets/application.scss (hunk 16 above)
Comment (domcleal) [reply app/assets/stylesheets/application.scss]: If you're intending to change the colour then I'd suggest either not doing it, using `<code>` tags around individual items in the list to better separate them, or extend the `code` CSS within the table to change the color/background properties.

File: app/helpers/templates_helper.rb (hunk 13 above)
Comment (lzap) [reply app/helpers/templates_helper.rb]: Sure.

File: app/views/provisioning_templates/_alerts.html.erb (hunk 14 above)
Comment (lzap) [reply app/views/provisioning_templates/_alerts.html.erb]: Ok I will move it and keep it as alert. My idea is to be bold enough so users get the idea quickly.

File: app/views/templates/_form.html.erb (hunk 15 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Not sure I understand but I will show it always, the warning box within the help tab will be only visible when safemode is on tho.

File: app/assets/stylesheets/application.scss (hunk 16 above)
Comment (lzap) [reply app/assets/stylesheets/application.scss]: Yeah I will try to come up with something better.

File: app/assets/stylesheets/application.scss (hunk 17 above)
Comment (lzap) [reply app/assets/stylesheets/application.scss]: Because I was hoping to retire without decent CSS knowledge. It does not work... :-D

File: app/views/templates/_form.html.erb (hunk 15 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Shall I keep "In templates only selected methods and variables are allowed" or do you think something better?

**Comment by lzap** [comment]
Pushed https://gist.github.com/e6ee45a7fcdf2d0aa342a0bd5c09effb

File: app/views/provisioning_templates/_alerts.html.erb (hunk 14 above)
Comment (domcleal) [reply app/views/provisioning_templates/_alerts.html.erb]: This is still on the first tab, shouldn't it be on the Help tab?

File: app/views/templates/_form.html.erb (hunk 15 above)
Comment (domcleal) [reply app/views/templates/_form.html.erb]: > This will need to be always displayed when moving the link to docs.

In the previous comment, I'd suggested moving the link to the useful macros etc to the Help tab, which you said you'd move. If it moved then the help tab would always need to be visible - which it now is, so sorted.

File: app/helpers/templates_helper.rb [hunk 18]
```diff
@@ -28,6 +28,28 @@ def infobox_functions_and_macros(docs_section)
...
+  def safemode_helpers
+    @@safemode_helpers ||= Foreman::Renderer::ALLOWED_HELPERS.sort.join(' ')
+  end
+
+  def safemode_variables
+    @@safemode_variables ||= Foreman::Renderer::ALLOWED_VARIABLES.sort.map{|x| "@#{x}"}.join(' ')
+  end
+
+  def infobox_safemode
+    alert(:class => 'alert-info', :header => '', :text => (_('In templates only selected methods and variables are allowed')))
```
Comment (domcleal) [inline app/helpers/templates_helper.rb]: > Shall I keep "In templates only selected methods and variables are allowed" or do you think something better?

I don't think it's a good alert as it doesn't really provide much information. It's probably better as text introducing the table, perhaps briefly stating that the list shows which methods may be called depending on the data type.

**Review by domcleal — CHANGES_REQUESTED** [review]
Per https://github.com/theforeman/foreman/pull/3744#discussion-diff-77809378, the link to useful macros etc should be on the Help tab now.

File: app/helpers/templates_helper.rb (hunk 18 above)
Comment (lzap) [reply app/helpers/templates_helper.rb]: Ok done, added examples block.

File: app/views/about/index.html.erb (hunk 6 above)
Comment (domcleal) [reply app/views/about/index.html.erb]: Still unused, should be removed.

File: app/views/templates/_form.html.erb [hunk 19]
```diff
@@ -84,6 +83,70 @@
       <% end %>
     </div>
 
+    <div class="tab-pane" id="help">
+      <div>
+        <%= _('Provisioning templates use ERB syntax') %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: Is this meant to be a heading? If it's meant to be part of the following sentence then it's missing a full stop, though doesn't make much sense as a full sentence.

File: app/views/templates/_form.html.erb [hunk 20]
```diff
@@ -84,6 +83,70 @@
...
+        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
+        <code class="transparent"><pre>
+<%%# this is a comment %>
+
+<%%# use Ruby for code flow %>
+<%% if true %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: use `-%>` for conditionals, most do

File: app/views/templates/_form.html.erb [hunk 21]
```diff
@@ -84,6 +83,70 @@
...
+<%%# host variable is available in most templates (Host::Managed table entry) %>
+<%%= @host %>
+
+<%%# associated operating system major version (Operatingsystem table entry) %>
+<%%= @host.operatingsystem.major %>
+
+<%%# associated hostgroup name (Hostgroup table entry) %>
+<%%= @host.hostgroup.name %>
+
+<%%# root password from associated hostgrup (without newline - note the trailing dash) %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: "host group", missing u and space

**Review by domcleal — CHANGES_REQUESTED** [review]
The "Check out some useful template functions and macros" link/alert is still on the first tab, but in an earlier update you said you'd move it to the Help tab. Instead, the new docs you've written are on a different tab to this link for more help.

File: app/views/templates/_form.html.erb (hunk 19 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: It ought to be a header, but I struggle to figure out something reasonable :-) Will do try then.

Edit: It was header not sentence. Whatever.

Comment (lzap) [reply app/views/templates/_form.html.erb]: How about:

```
        <%= _('Provisioning templates are generated using Embedded Ruby (ERB) language.') %>
        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
        <%= _('These are guaranteed to work via the safemode rendering, to ensure a template can do nothing harmful.') %>
        <%= _('With safemode disabled other macros may work, but are not supported at this time.') %>
        <%= _('To enable safemode, set "safemode_render" to "true" in Settings.') %>
```

Mostly coming from: http://projects.theforeman.org/projects/foreman/wiki/templatewriting

File: app/views/templates/_form.html.erb (hunk 20 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Sure.

File: app/views/templates/_form.html.erb (hunk 21 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Thanks.

File: app/views/templates/_form.html.erb (hunk 19 above)
Comment (domcleal) [reply app/views/templates/_form.html.erb]: It's OK, though the later sentences are quite short so it doesn't read too well to me. I'd also only suggest showing information about how to enable safemode if it's actually disabled - which is not the default, or leave it out.

Also I think the text in settings uses "safe mode" as two words.

**Comment by lzap** [comment]
Reworded, pushed.

File: app/views/templates/_form.html.erb [hunk 22]
```diff
@@ -84,6 +83,73 @@
       <% end %>
     </div>
 
+    <div class="tab-pane" id="help">
+      <div>
+        <%= _('Provisioning templates are generated using Embedded Ruby (ERB) language.') %>
+        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
+        <%= _('These are guaranteed to work via the safemode rendering, to ensure a template can do nothing harmful.') %>
+        <%= _('With safemode disabled other macros may work, but are not supported at this time.') %>
+        <%= _('To enable safemode, set "safemode_render" to "true" in Settings.') %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: Please see earlier comments at https://github.com/theforeman/foreman/pull/3744#discussion_r79589977

**Review by domcleal — CHANGES_REQUESTED** [review]

File: app/views/templates/_form.html.erb (hunk 19 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Sure, I will drop `With safemode disabled other macros may work, but are not supported at this time.` as it is not that relevant and only show the last sentence when appropriate. FYI the setting is named "Safemode rendering" but the description says "safe mode" - I will follow that. Pushing in a minute.

File: app/views/templates/_form.html.erb (hunk 9 above)
Comment (domcleal) [reply app/views/templates/_form.html.erb:30]: I don't think it's an improvement, the new position puts the snippet/family information directly between the template text and the template text upload field, which are related. I think the snippet/family information was more related to the name (and other general attributes), or even on another tab like provisioning templates.

File: app/views/templates/_form.html.erb [hunk 23]
```diff
@@ -54,6 +45,14 @@
...
+      <%= render "custom_tabs", :f => f if type == 'ptable' %>
+
+      <% if pxe_with_building_hosts?(@template) -%>
+        <% warning_text = (_("The template is associated to at least one host in build mode. To apply the change, disable and enable build mode on hosts to update the live templates or choose to %s their configuration from 'Select Action' menu") % link_to(_('recreate'), building_hosts_path(@template))).html_safe %>
+        <%= alert(:class => 'alert-warning', :header => 'Warning',
+                  :text => warning_text) %>
+      <% end -%>
+
       <% unless @template.locked? -%>
         <%= file_field_f f, :template, :class => "editor_file_source",:size => "col-md-10", :id => 'template_file',
```
Comment (domcleal) [inline app/views/templates/_form.html.erb:55]: To continue https://github.com/theforeman/foreman/pull/3744/files#r76795949, this field is related to the template edit box above - adding the other ptable fields here breaks the link between them.

File: app/views/templates/_form.html.erb [hunk 24]
```diff
@@ -84,6 +83,74 @@
       <% end %>
     </div>
 
+    <div class="tab-pane" id="help">
+      <div>
+        <%= _('Provisioning templates are generated using Embedded Ruby (ERB) language.') %>
+        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
+        <%= _('These are guaranteed to work via the safe mode rendering, to ensure a template can do nothing harmful.') %>
+        <% unless Setting[:safemode_render] -%>
+          <%= _('To enable safe mode, navigate to Settings page and flip Safemode rendering option.') %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: "flip" is rather informal, "enable" or "set the "Safemode rendering" option to "Yes"." may be better.

File: app/views/templates/_form.html.erb [hunk 25]
```diff
@@ -84,6 +83,74 @@
...
+<%%# global method random is listed in the table %>
+<%%= rand(9999) + 1 %>
+
+<%%# iterating through all network interaces of a host %>
+<%% @host.interfaces.each do |interface| -%>
+  NIC name: <%%= interface.name %>
+<%% end -%>
+        </pre></code>
+      </div>
+      <table class="table table-striped table-fixed">
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: missing `table_css_classes` (which obsoletes table, table-striped)

File: app/views/templates/_form.html.erb [hunk 26]
```diff
@@ -84,6 +83,74 @@
...
+<%%# print data from Ruby %>
+<%%= "some content from variable" %>
+
+<%%# host variable is available in most templates (Host::Managed table entry) %>
+<%%= @host %>
+
+<%%# associated operating system major version (Operatingsystem table entry) %>
+<%%= @host.operatingsystem.major %>
+
+<%%# associated hostgroup name (Hostgroup table entry) %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: "hostgroup" is inconsistent with the next description, so use "host group" in the first bit of the text, but leave "Hostgroup" in parentheses as that's the class name in the table

File: app/views/templates/_form.html.erb [hunk 27]
```diff
@@ -84,6 +83,74 @@
...
+        <% unless Setting[:safemode_render] -%>
+          <%= _('To enable safe mode, navigate to Settings page and flip Safemode rendering option.') %>
+        <% end -%>
+        <code class="transparent"><pre>
+<%%# this is a comment %>
+
+<%%# use Ruby for code flow %>
+<%% if true -%>
+  some content
+<%% end %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: also `-%>` here

File: app/views/templates/_form.html.erb [hunk 28]
```diff
@@ -84,6 +83,74 @@
...
 
+    <div class="tab-pane" id="help">
+      <div>
+        <%= _('Provisioning templates are generated using Embedded Ruby (ERB) language.') %>
+        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
+        <%= _('These are guaranteed to work via the safe mode rendering, to ensure a template can do nothing harmful.') %>
+        <% unless Setting[:safemode_render] -%>
+          <%= _('To enable safe mode, navigate to Settings page and flip Safemode rendering option.') %>
+        <% end -%>
+        <code class="transparent"><pre>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: Is `<code>` superfluous here? Pre should keep formatting, and also doesn't appear to need the CSS override.

File: app/views/templates/_form.html.erb [hunk 29]
```diff
@@ -84,6 +83,74 @@
       <% end %>
     </div>
 
+    <div class="tab-pane" id="help">
+      <div>
+        <%= _('Provisioning templates are generated using Embedded Ruby (ERB) language.') %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: "using _the_ Embedded [..]"?

**Review by domcleal — CHANGES_REQUESTED** [review]
Some inline comments there, but more generally, I'm concerned that the layout of the new code examples block and text you've added is poor - there's little space left between the elements (the text, to the pre block, to the table) and it looks rather thrown together.

Adding a paragraph break between the intro text and table, then a sub-heading above the table to describe what it is may make it look tidier.

> The "Check out some useful template functions and macros" link/alert is still on the first tab, but in an earlier update you said you'd move it to the Help tab. Instead, the new docs you've written are on a different tab to this link for more help.

This review comment from the last updates needs addressing still.

File: app/views/templates/_form.html.erb (hunk 23 above)
Comment (lzap) [reply app/views/templates/_form.html.erb:55]: I've reverted the whole form as it was before.

File: app/views/templates/_form.html.erb (hunk 28 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: I dropped `<code>` completely and redesigned the whole examples with PatternFly element.

File: app/views/templates/_form.html.erb (hunk 26 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: I removed this example completely.

**Comment by lzap** [comment]
Pushed the all new UI. Looks much better now.

File: app/views/templates/_form.html.erb [hunk 30]
```diff
@@ -84,6 +81,74 @@
...
+        <%= _('Provisioning templates are generated using the Embedded Ruby (ERB) language.') %>
+        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
+        <%= _('These are guaranteed to work via the safe mode rendering, to ensure a template can do nothing harmful.') %>
+        <% unless Setting[:safemode_render] -%>
+          <%= _('To enable safe mode, navigate to Settings page and enable Safemode rendering option.') %>
+        <% end -%>
+        <%= _('For more info %s.').html_safe % link_to(_('visit our documentation'), documentation_url(controller.documentation_archor)) %>
+        <h2><%= _('Examples') %></h2>
+        <div class="panel-group" id="accordion-markup">
+          <%= render "help_accordion", :title => 'Comments', :contents => (capture { %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: Please extract the `:title` strings for i18n in all of these accordions.

File: app/views/templates/_form.html.erb [hunk 31]
```diff
@@ -84,6 +81,74 @@
...
+    <div class="tab-pane" id="help">
+      <div>
+        <h1><%= _('Template syntax') %></h1>
+        <%= _('Provisioning templates are generated using the Embedded Ruby (ERB) language.') %>
+        <%= _('The following table and example snippet shows global methods, variables and class methods and how to use them.') %>
+        <%= _('These are guaranteed to work via the safe mode rendering, to ensure a template can do nothing harmful.') %>
+        <% unless Setting[:safemode_render] -%>
+          <%= _('To enable safe mode, navigate to Settings page and enable Safemode rendering option.') %>
+        <% end -%>
+        <%= _('For more info %s.').html_safe % link_to(_('visit our documentation'), documentation_url(controller.documentation_archor)) %>
```
Comment (domcleal) [inline app/views/templates/_form.html.erb]: I'd suggest just linking the entire sentence, it'll be far easier to translate - translators won't know that these two strings are related in any way from Transifex. `_("Visit our documentation for more information.")` or similar and link it all.

File: app/controllers/provisioning_templates_controller.rb [hunk 32]
```diff
@@ -1,9 +1,14 @@
 class ProvisioningTemplatesController < TemplatesController
   include Foreman::Controller::Parameters::ProvisioningTemplate
+  helper_method :documentation_archor
```
Comment (domcleal) [inline app/controllers/provisioning_templates_controller.rb]: Typo: "anchor"

File: app/views/templates/_help_accordion.html.erb [hunk 33]
```diff
@@ -0,0 +1,16 @@
+          <div class="panel panel-default">
```
Comment (domcleal) [inline app/views/templates/_help_accordion.html.erb]: Nitpick, lose the indentation, I don't think we normally indent fragments to match where they're called from, it's easier to edit (even if it renders without indentation).

**Review by domcleal — CHANGES_REQUESTED** [review]
Thanks, that's a huge improvement.

File: app/views/templates/_form.html.erb (hunk 31 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Sure.

File: app/views/templates/_form.html.erb (hunk 30 above)
Comment (lzap) [reply app/views/templates/_form.html.erb]: Yeah I was thinking - when the rest (inline comments in the examples) are not translated - not to do it. Whatever.

File: app/views/templates/_help_accordion.html.erb (hunk 33 above)
Comment (lzap) [reply app/views/templates/_help_accordion.html.erb]: Sure, all comments done in https://gist.github.com/2ae30894d7a8157507650aa5e696e1ea

**Comment by domcleal** [comment]
[test foreman]

**Review by domcleal — APPROVED** [review]

**Comment by domcleal** [comment]
Merged as 7907984, thanks @lzap.
